### PR TITLE
Feature/specification

### PR DESCRIPTION
**Pull request by cvarjao**
Previews:
- [readme.md](https://github.com/bcgov/nrs-cmdb/blob/feature/specification/spec/readme.md)
- [spec](https://github.com/bcgov/nrs-cmdb/blob/feature/specification/spec/)

TODO:
- [ ] Port assignment/reservation (e.g.: Tomcat)
- [ ] deploymentSpec vs runtimeSpec ... do we need them?
- [ ] requires format (e.: ojdbc.url, ojdbc.oid.url). multiple formats?
- [x] support selector equivalent of kubernetes `matchLabels` or `matchExpressions`

### CODE REVIEW

File: spec/readme.md [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+  * if `ArtifactVersion` node of type `name` does not exist, create a `ArtifactVersion` node for the `System`
+  * Create `DeploymentSpec` node for the `System`, `ArtifactVersion` and `environment`
+  * Process `runtimeSpec.provides`
+    * if `Interface` node matching `interface` does not exist, creates a `Interface` node for the `System`
+    * Create of Update a `Component` node for the `DeploymentSpec` node withe `Interface` node as the `interface` attribute
+  * Process `runtimeSpec.requires`
+  * Process `deploymentSpec`
+1. Parse _Deployment Specification Plan_, extract parameters/arguments and execute deployment.
+1. POST the _Deployment Specification Plan_ to `/deployments/<plan id>/finish` indicating if it was successful or not.
+
```
Comment (GeorgeWalker) [inline spec/readme.md:25]: Do we need to post the deployment specification plan again?  If so, would the finish update the data store if the object has changed?

File: spec/readme.md [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+## DeploymentSpec
+key/value map where the value is of type [Selector](#selector)
+
+## Selector
+| Field | Type | Default |Description |
+| :--- | :--- | :--- | :--- |
+| selector | array of [MatchExpression](#MatchExpression) | `[{"key":"environment", "operator":"In", "value":["${input.environemnt}"]}]` | |
+| quantifier | String | `1` | `+` (zero or one), `*` (zero or more), `?` (one or more), `n` (exact `n` times) |
+| attributes | array of string | `[]` | Define wicth attributes to expanded/provided. Empty array means `all`.|
```
Comment (GeorgeWalker) [inline spec/readme.md]: typo - should be Define which attributes to be expanded/provided

File: spec/readme.md [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+| :--- | :--- | :--- | :--- |
+| key | string | | |
+| operator | enum (string) | `"In"` | `"In"`, `"NotIn"` |
+| values | array of [stringTemplate](#string-template) | `[]` | |
+
+## RuntimeSpec
+| Field | Type | Description |
+| :--- | :--- | :--- |
+| provides | array of [Provide](#provide) | |
+| requires | array of [Require](#pequire) | |
```
Comment (GeorgeWalker) [inline spec/readme.md]: Did you mean #require here?

**Comment by cvarjao** [comment]
@GeorgeWalker , regarding the questions/TODO in the PR. Any thoughts? Specially on the splitting of runtime vs deployment spec. I've been thinking we can have one only and add a "scope" property that defined if it is runtime (default)/deployment. and on how to provide formatted attributes.

File: spec/readme.md (hunk 1 above)
Comment (cvarjao) [reply spec/readme.md:25]: @GeorgeWalker, I was thinking the deployment process could provide some extra metadata (mostly attributes). But I am good with defining (or at least starting with) as one-way.

File: spec/example-artifacts.json [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+  {
+    "system": "OFM",
+    "name": "com.oracle.ofm",
+    "shortName": "Oracle Fusion Middleware",
+    "description": "Oracle's complete family of application infrastructure products—from the #1 Java application server to SOA and enterprise portals—are integrated with Oracle Applications and technologies to speed implementation and lower the cost of management and change. Best-of-breed offerings and unique hot-pluggable capabilities provide a foundation for innovation and extend the business value of existing investments.",
+    "url": "",
+    "vendor": "Oracle",
+    "vendorContact": "support@oracle.com",
+    "version": "11.1.1",
+    "deploymentSpec":{
```
Comment (GeorgeWalker) [inline spec/example-artifacts.json]: deploymentSpec and runtimeSpec appear redundant.  In my opinion we could simply have a host reference, and a provides reference (which would be a vector, perhaps of export objects).

**Review by cvarjao — COMMENTED** [review]
@GeorgeWalker, what do you think about this change? I still need to update the spec docs, I just updated the example file.

**Comment by GeorgeWalker** [comment]
The most recent change looks good @cvarjao.  I'll update the demo to reflect it.

**Pull request merged**